src/kernel/bitmap.zig

```diff
@@ -0,0 +1,228 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const testing = std.testing;
+
+pub fn Bitmap(comptime BitmapType: type) type {
+    return struct {
+        /// The possible errors thrown by bitmap functions
+        pub const BitmapError = error{
+            /// The address given was outside the region covered by a bitmap
+            OutOfBounds,
+        };
+
+        const Self = @This();
+
+        /// The number of entries that one bitmap type can hold. Evaluates to the number of bits the type has
+        pub const ENTRIES_PER_BITMAP: u32 = std.meta.bitCount(BitmapType);
+
+        /// The value that a full bitmap will have
+        pub const BITMAP_FULL = std.math.maxInt(BitmapType);
+
+        /// The type of an index into a bitmap entry. The smallest integer needed to represent all bit positions in the bitmap entry type
+        pub const IndexType = @Type(builtin.TypeInfo{ .Int = builtin.TypeInfo.Int{ .is_signed = false, .bits = std.math.log2(std.meta.bitCount(BitmapType)) } });
+
+        num_bitmaps: u32,
+        num_entries: u32,
+        bitmaps: []BitmapType,
+        num_free_entries: u32,
+
+        pub fn init(num_entries: u32, allocator: *std.mem.Allocator) !Self {
+            const num = num_entries / ENTRIES_PER_BITMAP + @boolToInt(num_entries % ENTRIES_PER_BITMAP != 0);
+            const self = Self{
+                .num_bitmaps = num,
+                .num_entries = num_entries,
+                .bitmaps = try allocator.alloc(BitmapType, num),
+                .num_free_entries = num_entries,
+            };
+            for (self.bitmaps) |*bmp| {
+                bmp.* = 0;
+            }
+            return self;
+        }
+
+        ///
+        /// Set an entry within a bitmap as occupied.
+        ///
+        /// Arguments:
+        ///     OUT bmp: *Bitmap - The bitmap to modify.
+        ///     IN idx: BitmapIndex - The index within the bitmap to set.
+        ///
+        pub fn setEntry(self: *Self, idx: u32) BitmapError!void {
+            if (idx >= self.num_entries) return BitmapError.OutOfBounds;
+            if (!try self.isSet(idx)) {
+                const bit = @as(BitmapType, 1) << @intCast(IndexType, idx % ENTRIES_PER_BITMAP);
```

Comment:
This could be made into a function as it is the same as the `clearEntry`. Plus this is different to the `clearEntry`